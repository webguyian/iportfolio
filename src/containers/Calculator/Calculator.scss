.calculator-app {
  display: flex;
  flex-flow: column wrap;

  &-output {
    @include type-number(9vw);
    display: flex;
    flex-flow: column wrap;
    justify-content: flex-end;
    width: 365px;
    height: 300px;
    padding: 0 10px 10px;
    line-height: 1;
    text-align: right;
    white-space: nowrap;
    overflow: hidden;

    &--8 {
      font-size: 8vw;
    }

    &--9 {
      font-size: 7vw;
    }
  }

  &-input {
    @include no-outline;
  }

  &-row {
    display: flex;
    justify-content: space-evenly;
    width: 100%;

    > .ui-btn {
      width: 80px;
      height: 80px;
      font-size: 33px;
      line-height: 1;

      &:last-child {
        background-color: $orange;
        font-size: 40px;
      }
    }

    &--1 {
      > .ui-btn {
        background-color: $gray;
      }
    }

    &--5 {
      > .ui-btn {
        &:first-child {
          width: 170px;
          padding-left: 30px;
          border-radius: 40px;
          text-align: left;
        }
      }
    }
  }
}
