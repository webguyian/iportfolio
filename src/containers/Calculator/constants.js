export const add = '+';
export const allClear = 'AC';
export const backspace = 'Backspace';
export const clear = 'C';
export const decimal = '.';
export const divide = '/';
export const divideAlt = '\u00f7';
export const enter = 'Enter';
export const equals = '=';
export const multiply = '*';
export const multiplyAlt = '\u00d7';
export const percent = '%';
export const positiveNegative = '+/-';
export const subtract = '-';
export const zero = '0';

export const controls = {
  add,
  allClear,
  clear,
  decimal,
  divide,
  divideAlt,
  equals,
  multiply,
  multiplyAlt,
  percent,
  positiveNegative,
  subtract
};

export const allOperators = [
  add,
  allClear,
  clear,
  decimal,
  divide,
  divideAlt,
  equals,
  multiply,
  multiplyAlt,
  percent,
  positiveNegative,
  subtract
];

export const operators = [
  add,
  allClear,
  clear,
  divide,
  divideAlt,
  multiply,
  multiplyAlt,
  positiveNegative,
  subtract
];

export const controlLabels = {
  [allClear]: 'All clear',
  [clear]: 'Clear',
  [positiveNegative]: 'Positive/negative',
  [percent]: 'Percent',
  [multiplyAlt]: 'Multiply',
  [divideAlt]: 'Divide',
  [add]: 'Add',
  [subtract]: 'Subtract',
  [equals]: 'Equals',
  [decimal]: 'Decimal'
};

export const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];

export const keys = {
  backspace,
  enter
};

export const operatorKeys = [backspace, enter]
  .concat(...allOperators)
  .concat(numbers);
