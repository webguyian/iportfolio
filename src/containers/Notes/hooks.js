import { useEffect, useRef, useState } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useRefFocus = (condition = true, initial = null) => {
  const ref = useRef(initial);

  useEffect(() => {
    const element = ref && ref.current;

    if (condition && element) {
      // Add focus to element
      element.focus();
    }
  }, [condition, ref]);

  return ref;
};

export const useSwipeOffset = (initial = 0) => {
  const [offset, setOffset] = useState(initial);
  const maxOffset = -30;

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Down') {
      // Exit early if not down swipe
      return false;
    }

    setOffset(eventInfo.deltaY);
  };

  const handleSwipeUp = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedUp: handleSwipeUp,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset < maxOffset, setOffset];
};
