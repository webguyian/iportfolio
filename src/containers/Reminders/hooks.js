import { useEffect, useState, useRef } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useLocalStorage = (key, initialValue) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored json or if none return initialValue
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      // If error return initialValue
      return initialValue;
    }
  };
  const [storedValue, setStoredValue] = useState(getValue);
  const setValue = value => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore =
        value instanceof Function ? value(storedValue) : value;

      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      window.localStorage.setItem(key, JSON.stringify(valueToStore));
    } catch (error) {
      // Drop error
    }
  };

  return [storedValue, setValue, getValue];
};

export const useRefFocus = (node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element) {
      // Add focus to element
      element.focus();
    }
  }, [ref]);

  return ref;
};

export const useRefControlledFocus = (focused, node = null) => {
  const ref = useRef(node);

  useEffect(() => {
    const element = ref && ref.current;

    if (element && focused) {
      element.focus();
    }
  }, [focused]);

  return ref;
};

export const useReminders = (reminders, setReminders) => {
  const remindersRef = useRef(reminders);

  useEffect(() => {
    // Update reminders in ref
    remindersRef.current = reminders;
  });

  useEffect(() => {
    return () => {
      const { current } = remindersRef;
      const updatedReminders = current.filter(reminder =>
        reminder.value.trim()
      );

      // Trigger callback without empty reminders
      setReminders(updatedReminders);
    };
  }, [remindersRef]);

  return remindersRef;
};

export const useSwipeOffset = (callback, initial = 0) => {
  const [offset, setOffset] = useState(initial);

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Left') {
      // Exit early if not left swipe
      return false;
    }

    setOffset(eventInfo.deltaX);
  };

  const handleSwipeLeft = eventInfo => {
    const { deltaX } = eventInfo;

    if (deltaX > 80) {
      callback(eventInfo);
    }
  };

  const handleSwipeRight = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedLeft: handleSwipeLeft,
    onSwipedRight: handleSwipeRight,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset, setOffset];
};
