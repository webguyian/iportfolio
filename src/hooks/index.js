import { useEffect, useState } from 'react';

export const useBreakpoint = breakpoint => {
  const [matchesBreakpoint, setBreakpoint] = useState(false);
  const mediaQuery = `(max-width: ${breakpoint}px)`;

  const handleResize = () => {
    const matches = window.matchMedia(mediaQuery).matches;

    setBreakpoint(matches);
  };

  useEffect(() => {
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [breakpoint]);

  return matchesBreakpoint;
};
