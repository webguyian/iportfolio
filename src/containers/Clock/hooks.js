import { useEffect, useRef, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { initialValues } from './constants';
import { getExpiration, getExpirationValues, getSeconds } from './helpers';

export const useInitialRoute = (match, path) => {
  const history = useHistory();

  useEffect(() => {
    if (match.isExact) {
      history.push(path);
    }
  }, []);

  return history;
};

export const useStorage = (key, currentValue, deleteKey, keepUpdating) => {
  const getValue = () => {
    try {
      // Get item from local storage with key
      const item = window.localStorage.getItem(key);

      // Parse stored JSON or if none return undefined
      return item ? JSON.parse(item) : undefined;
    } catch (error) {
      // If error return undefined
      return undefined;
    }
  };

  const value = getValue();
  const localRef = useRef(value);
  const [storedValue, setStoredValue] = useState(value);

  const setValue = nextValue => {
    try {
      // Save to state and local storage
      setStoredValue(nextValue);

      if (deleteKey && !nextValue[deleteKey]) {
        // Remove item from local storage
        window.localStorage.removeItem(key);
      } else {
        // Update local storage and include timestamp
        nextValue.timestamp = Date.now();
        window.localStorage.setItem(key, JSON.stringify(nextValue));
      }
    } catch (error) {
      // Drop error
    }
  };

  useEffect(() => {
    // Update local value
    localRef.current = currentValue || storedValue;

    if (keepUpdating) {
      // Set update value in local storage
      setValue(localRef.current);
    }
  }, [currentValue, storedValue]);

  useEffect(() => {
    return () => {
      if (key && localRef.current) {
        setValue(localRef.current);
      }
    };
  }, []);

  return localRef.current;
};

export const useLaps = (initialLaps, timer, timerRunning) => {
  const [laps, setLaps] = useState(initialLaps);
  const hasLaps = laps.some(lap => lap > 0);
  const storage = useStorage(
    'stopwatch',
    {
      laps,
      timer,
      timerRunning
    },
    'timer'
  );

  const updateLaps = () => {
    if (timerRunning) {
      const hasEmptyLaps = laps.some(lap => lap === 0);

      if (hasEmptyLaps) {
        setLaps(updatedLaps => [
          0,
          ...updatedLaps.slice(0, updatedLaps.length - 1)
        ]);
      } else {
        setLaps(updatedLaps => [0, ...updatedLaps]);
      }
    } else {
      setLaps(initialLaps);
    }
  };

  useEffect(() => {
    if (storage && storage.laps) {
      setLaps(storage.laps);
    }
  }, []);

  useEffect(() => {
    if (!timerRunning || !timer) {
      return;
    }

    const lapLength = laps.length;

    if (!hasLaps && timer > 10) {
      setLaps(updatedLaps => [timer, ...updatedLaps]);
    }

    if (lapLength > 5) {
      setLaps(updatedLaps => {
        const truncatedLaps = updatedLaps.slice(1, lapLength);
        const offset = truncatedLaps.reduce((a, b) => a + b, 0);

        return [timer - offset, ...truncatedLaps];
      });
    }
  }, [timer, timerRunning, hasLaps]);

  return [laps, updateLaps, hasLaps];
};

export const useStopwatch = (startTime = 0, start = false) => {
  const [timerStarted, startTimer] = useState(start);
  const [interval, setLocalInterval] = useState(null);
  const [offset, setOffset] = useState(0);
  const timer = useRef(startTime);
  const storage = useStorage('stopwatch');
  const toggleTimer = () => startTimer(!timerStarted);
  const resetTimer = () => {
    timer.current = 0;
    return timer.current;
  };
  const getUpdatedTime = cache => {
    const now = Date.now();
    const diff = now - cache.timestamp;

    return cache.timer + diff;
  };

  useEffect(() => {
    if (storage && storage.timer) {
      // Get updated time from local storage
      if (storage.timerRunning) {
        timer.current = getUpdatedTime(storage);
        toggleTimer();
      } else {
        timer.current = storage.timer;
      }
    }
  }, []);

  useEffect(() => {
    let localOffset = offset;

    if (timerStarted && !interval) {
      localOffset = Date.now();

      const getDelta = () => {
        const now = Date.now();
        const delta = now - localOffset;

        localOffset = now;

        setOffset(now);
        return delta;
      };

      const localInterval = window.setInterval(() => {
        timer.current += getDelta();
      }, 1);

      setLocalInterval(localInterval);
    }

    if (!timerStarted && interval) {
      window.clearInterval(interval);
      setLocalInterval(null);
    }
  }, [timerStarted, timer, interval]);

  return [timer.current, timerStarted, toggleTimer, resetTimer];
};

export const useDuration = () => {
  const [values, setValues] = useState(null);
  const storage = useStorage('duration', values, null, true);
  const setDuration = (key, value) => {
    setValues({
      ...values,
      [key]: value
    });
  };
  const cachedValues = storage || values;
  const duration = cachedValues && getSeconds(cachedValues);

  useEffect(() => {
    if (storage) {
      // Update state with values from local storage
      setValues(storage);
    } else {
      // Set default initial values
      setValues(initialValues);
    }
  }, []);

  return [duration, setDuration, cachedValues, setValues];
};

export const useCountdown = (h = 0, m = 0, s = 0) => {
  const [seconds, setSeconds] = useState(s);
  const [minutes, setMinutes] = useState(m);
  const [hours, setHours] = useState(h);
  const allSeconds = getSeconds({ hours, minutes, seconds });

  const subtractHour = () => {
    setHours(prevHours => {
      if (prevHours === 0) {
        return 23;
      } else if (prevHours > 0) {
        return prevHours - 1;
      }
      return 0;
    });
  };

  const subtractMinute = () => {
    setMinutes(prevMinutes => {
      if (prevMinutes === 0) {
        subtractHour();
        return 59;
      } else if (prevMinutes > 0) {
        return prevMinutes - 1;
      }
      return 0;
    });
  };

  const subtractSecond = () => {
    setSeconds(prevSeconds => {
      if (prevSeconds === 0) {
        subtractMinute();
        return 59;
      } else if (prevSeconds > 0) {
        return prevSeconds - 1;
      }
      return 0;
    });
  };

  const setCountdown = values => {
    if (!values) {
      setHours(0);
      setMinutes(0);
      setSeconds(0);
      return;
    }

    setHours(values.hours);
    setMinutes(values.minutes);
    setSeconds(values.seconds);
  };

  return [
    {
      allSeconds,
      hours,
      minutes,
      seconds
    },
    setCountdown,
    subtractSecond
  ];
};

export const useTimer = duration => {
  const [started, setStarted] = useState(false);
  const [running, setRunning] = useState(false);
  const [countdown, setCountdown, startCountdown] = useCountdown();
  const intervalRef = useRef();
  const storage = useStorage(
    'timer',
    {
      ...countdown,
      duration,
      running,
      started
    },
    'allSeconds'
  );

  const pause = () => {
    if (intervalRef.current) {
      setRunning(() => false);
      clearInterval(intervalRef.current);
      intervalRef.current = undefined;
    }
  };

  const reset = () => {
    pause();
    setCountdown(null);
    setStarted(false);
  };

  const calculateExpiration = expiration => {
    const distance = expiration - Date.now();
    const expirationValues = getExpirationValues(distance);

    if (countdown.allSeconds < 0) {
      reset();
    } else {
      setCountdown(expirationValues);
    }

    return expirationValues;
  };

  const start = expiration => {
    if (expiration && !intervalRef.current) {
      calculateExpiration(expiration);
    }
  };

  const resume = () => {
    if (!intervalRef.current) {
      setStarted(() => true);
      setRunning(() => true);
      intervalRef.current = setInterval(startCountdown, 1000);
    }
  };

  const restart = () => {
    const expiration = getExpiration(duration);

    reset();
    start(expiration);
  };

  const toggle = () => {
    if (!intervalRef.current) {
      resume();
    } else {
      pause();
    }
  };

  useEffect(() => {
    const expiration = getExpiration(duration, storage);

    start(expiration);

    if (storage) {
      if (storage.running) {
        resume();
      } else if (storage.started) {
        setStarted(true);
      }

      if (duration !== storage.duration) {
        // Remove timer from local storage
        window.localStorage.removeItem('timer');
      }
    }

    return pause;
  }, [duration]);

  return {
    ...countdown,
    duration,
    start,
    started,
    pause,
    reset,
    resume,
    restart,
    running,
    toggle
  };
};

export const useRadialOffset = (timer, radius) => {
  const circumference = (2 * radius * 22) / 7;
  const { allSeconds, duration } = timer;
  const [initialSeconds] = useState(allSeconds);
  const [offset, setOffset] = useState(0);

  useEffect(() => {
    if (!offset) {
      const initialOffset = (allSeconds / duration) * circumference;
      const updatedOffset = circumference - Math.round(initialOffset);

      if (updatedOffset === circumference || updatedOffset < 0) {
        // Exit early
        return;
      }

      setOffset(updatedOffset);
    }
  }, [allSeconds, duration, offset]);

  return [offset, initialSeconds];
};
