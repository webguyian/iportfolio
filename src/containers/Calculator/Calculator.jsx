import React from 'react';
import classNames from 'classnames';

import RoundedButton from 'components/Button/RoundedButton';
import Text from 'components/Text/Text';
import { controls, controlLabels, keys, operatorKeys } from './constants';
import { getSymbols } from './helpers';
import { useCalculator, useCalculatorRef } from './hooks';

const Calculator = () => {
  const baseClass = 'calculator-app';
  const rowClass = `${baseClass}-row`;
  const calculatorEl = useCalculatorRef(null);
  const [display, setKey] = useCalculator('0');
  const outputClass = `${baseClass}-output`;
  const lengthClass = `${outputClass}--${display.length}`;
  const symbols = getSymbols(display);

  const handleKeyPress = ({ key }) => {
    if (!operatorKeys.includes(key)) {
      return false;
    }

    if (key === keys.backspace) {
      return setKey(controls.clear);
    }

    if (key === keys.enter) {
      return setKey(controls.equals);
    }

    return setKey(key);
  };

  const handleClickOutput = () => {
    // Add focus to wrapper element
    calculatorEl.current.focus();

    if (navigator.clipboard) {
      // Copy display text to clipboard
      navigator.clipboard.writeText(display);
    }
  };

  return (
    <div className={baseClass}>
      <output
        className={classNames(outputClass, lengthClass)}
        onClick={handleClickOutput}
      >
        {display}
      </output>
      <div
        className={`${baseClass}-input`}
        ref={calculatorEl}
        tabIndex={-1}
        onKeyUp={handleKeyPress}
      >
        {symbols.map((row, rowIndex) => (
          <div
            key={rowIndex}
            className={`${rowClass} ${rowClass}--${rowIndex + 1}`}
          >
            {row.map((symbol, symbolIndex) => (
              <RoundedButton
                key={symbolIndex}
                onClick={setKey.bind(null, symbol)}
              >
                {symbol}
                <Text type="accessible">{controlLabels[symbol]}</Text>
              </RoundedButton>
            ))}
          </div>
        ))}
      </div>
    </div>
  );
};

export default Calculator;
