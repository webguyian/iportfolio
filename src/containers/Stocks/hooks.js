import { useEffect, useState } from 'react';

import { useStorageCache } from 'hooks';
import { API_EXCHANGE, API_TOKEN } from 'containers/Stocks/constants';
import {
  getCandlestickEndpoint,
  getCandlestickEndpoints,
  getStockEndpoint,
  getStockNewsEndpoint,
  isExpired,
  isExpiredNews,
  updateChartData,
  rehydrateChartData
} from 'containers/Stocks/helpers';

export const useFetch = endpoint => {
  const [data, setData] = useState(null);
  const fetchData = async () => {
    const response = await fetch(endpoint);
    const result = await response.json();

    setData(result);
  };

  useEffect(() => {
    if (!endpoint) {
      return;
    }

    fetchData();
  }, [endpoint]);

  return data;
};

export const useFetchAll = urls => {
  const [responses, setResponses] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const results = await Promise.all(
        urls.map(async url => {
          const response = await fetch(url);

          return response.json();
        })
      );

      setResponses(results);
    };

    if (urls.length) {
      fetchData();
    }
  }, [urls]);

  return responses;
};

export const useSearch = () => {
  const [value, setValue] = useState('');
  const [searchTerm, setSearchTerm] = useState('');
  const [searchFocus, setSearchFocus] = useState(false);
  const hasSearch = Boolean(searchFocus || searchTerm.length);

  const handleCancel = () => {
    setValue('');
    setSearchTerm('');
    setSearchFocus(false);
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const handleFocus = () => {
    setSearchFocus(focus => !focus);
  };

  const searchHandlers = {
    onBlur: handleFocus,
    onCancel: handleCancel,
    onChange: handleChange,
    onFocus: handleFocus,
    value
  };

  useEffect(() => {
    if (value) {
      setSearchTerm(value.toUpperCase().trim());
    } else {
      setSearchTerm('');
    }
  }, [value]);

  return [searchTerm, searchHandlers, hasSearch];
};

export const useStockSearch = (allStocks, searchTerm) => {
  const [filteredStocks, setFilteredStocks] = useState(allStocks);

  useEffect(() => {
    if (searchTerm.length > 1) {
      const filtered = allStocks.filter(stock => {
        return (
          stock.symbol.indexOf(searchTerm) === 0 ||
          stock.description.indexOf(searchTerm) === 0
        );
      });

      setFilteredStocks(filtered);
    } else if (!searchTerm) {
      setFilteredStocks(allStocks);
    }
  }, [searchTerm]);

  return filteredStocks;
};

export const useStocks = initialStocks => {
  const [endpoints, setEndpoints] = useState([]);
  const [stocks, setStocks] = useState([]);
  const [allStocks, setAllStocks] = useState([]);
  const [activeStocks, setActiveStocks] = useState([]);
  const exchangeResponse = useFetch(API_EXCHANGE);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stocks',
    { stocks },
    stockData => !stockData.stocks.length
  );

  useEffect(() => {
    const cachedStocks = cache && cache.stocks;

    if (cachedStocks && cachedStocks.length) {
      setStocks(cachedStocks);
      setActiveStocks(cachedStocks);
    } else {
      setStocks(initialStocks);
      setActiveStocks(initialStocks);
    }
  }, []);

  useEffect(() => {
    if (!exchangeResponse) {
      // Exit early if no response
      return;
    }

    const symbols = activeStocks.map(stock => stock.symbol);
    const currentStocks = exchangeResponse.filter(stock =>
      symbols.includes(stock.symbol)
    );

    setEndpoints(currentStocks.map(getStockEndpoint));
    setAllStocks(exchangeResponse);
  }, [exchangeResponse]);

  useEffect(() => {
    if (activeStocks.length) {
      setEndpoints(activeStocks.map(getStockEndpoint));
    }
  }, [activeStocks]);

  useEffect(() => {
    if (!response.length || !activeStocks.length) {
      return;
    }

    const updatedStocks = response.map((result, index) => {
      const stock = activeStocks[index];

      return {
        ...result,
        ...stock,
        price: result.c,
        previousPrice: result.pc
      };
    });

    setStocks(updatedStocks);
  }, [response, activeStocks]);

  return [stocks, setStocks, allStocks];
};

export const useStockDetail = () => {
  const [endpoint, setEndpoint] = useState('');
  const [activeStock, setActiveStock] = useState(null);
  const [stock, setStockDetail] = useState(activeStock);
  const response = useFetch(endpoint);
  const getStock = symbol =>
    `https://finnhub.io/api/v1/quote?symbol=${symbol}&token=${API_TOKEN}`;

  useEffect(() => {
    setStockDetail(activeStock);

    if (!activeStock) {
      return;
    }

    if (!activeStock.price) {
      setEndpoint(getStock(activeStock.symbol));
    }
  }, [activeStock]);

  useEffect(() => {
    if (response) {
      setStockDetail({
        ...activeStock,
        ...response,
        price: response.c,
        previousPrice: response.pc
      });
      setEndpoint('');
    }
  }, [response]);

  return [stock, setActiveStock];
};

export const useStockNews = category => {
  const [endpoint, setEndpoint] = useState('');
  const [news, setNews] = useState([]);
  const response = useFetch(endpoint);
  const cache = useStorageCache(
    'stock-news',
    { news: response },
    res => !res.news.length
  );

  useEffect(() => {
    const cachedNews = cache && cache.news;

    if (cachedNews && cachedNews.length) {
      setNews(cachedNews);
      const expired = isExpiredNews(cachedNews[0].datetime);

      if (!expired && !category) {
        // Exit early when cache is not expired and no category
        return;
      }
    }

    setEndpoint(getStockNewsEndpoint(category));
  }, [category]);

  useEffect(() => {
    if (response && response.length) {
      setNews(response);
    }
  }, [response]);

  return news;
};

export const useStockChart = (symbol, range) => {
  const [endpoint, setEndpoint] = useState();
  const [updatedResponse, setResponse] = useState(null);
  const response = useFetch(endpoint);

  useEffect(() => {
    const updatedEndpoint = getCandlestickEndpoint(symbol, range);

    setEndpoint(updatedEndpoint);
  }, [symbol, range]);

  useEffect(() => {
    if (response && response.s === 'ok') {
      const data = updateChartData(response, range);

      setResponse(data);
    }
  }, [response]);

  return updatedResponse;
};

export const useStockTicker = (stocks, range = '1D') => {
  const [endpoints, setEndpoints] = useState([]);
  const [updatedResponse, setResponse] = useState([]);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stock-charts',
    { charts: updatedResponse },
    res => !res.charts.length
  );

  useEffect(() => {
    const cachedCharts = cache && cache.charts;

    if (cachedCharts && cachedCharts.length) {
      const updated = cachedCharts.map(rehydrateChartData);
      const [head] = updated;
      const expired = isExpired(head.chartData.endDate);

      setResponse(updated);

      if (!expired) {
        // Exit early when cache is not expired
        return;
      }
    }

    // Trigger fetch for candlestick data
    setEndpoints(getCandlestickEndpoints(stocks, range));
  }, [stocks.length]);

  useEffect(() => {
    if (response.length) {
      const updated = response.map((res, index) => {
        if (res.s !== 'ok') {
          return null;
        }

        const stock = stocks[index];
        const chartData = updateChartData(res, range);

        return {
          ...stock,
          chartData
        };
      });

      if (updated[0] !== null) {
        setResponse(updated);
      }
    }
  }, [response]);

  return updatedResponse;
};
