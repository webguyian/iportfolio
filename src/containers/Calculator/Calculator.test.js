import React from 'react';
import { act, create } from 'react-test-renderer';

import Calculator from './Calculator';

import * as hooks from 'modules/calculator/hooks';

describe('<Calculator />', () => {
  const originalUseCalculator = hooks.useCalculator;

  hooks.useCalculator = jest.fn();
  const setKey = jest.fn();
  const calculatorEl = {
    focus: jest.fn()
  };
  const createNodeMock = element => {
    return element.props.tabIndex === -1 ? calculatorEl : null;
  };

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset to original implementation before each test
    hooks.useCalculator.mockImplementation(originalUseCalculator);
  });

  it('renders correctly', () => {
    const component = create(<Calculator />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('handles output clicks', () => {
    const component = create(<Calculator />, {
      createNodeMock
    }).root;
    const output = component.findByType('output');

    expect(calculatorEl.focus).not.toHaveBeenCalled();
    output.props.onClick();
    expect(calculatorEl.focus).toHaveBeenCalled();
  });

  it('handles output clicks with clipboard', () => {
    Object.defineProperty(window.navigator, 'clipboard', {
      value: { writeText: jest.fn() }
    });

    const component = create(<Calculator />, {
      createNodeMock
    }).root;
    const output = component.findByType('output');

    calculatorEl.focus.mockClear();

    expect(calculatorEl.focus).not.toHaveBeenCalled();
    output.props.onClick();
    expect(calculatorEl.focus).toHaveBeenCalled();
    expect(navigator.clipboard.writeText).toHaveBeenCalledWith('0');
  });

  it('handles keypress for unknown key', () => {
    const component = create(<Calculator />, {
      createNodeMock
    }).root;
    const input = component.findByProps({ tabIndex: -1 });
    const mockEvent = {
      key: 'unknown'
    };

    expect(input.props.onKeyUp(mockEvent)).toBe(false);
  });

  it('handles keypress for Backspace key', () => {
    const component = create(<Calculator />, {
      createNodeMock
    }).root;
    const input = component.findByProps({ tabIndex: -1 });
    const mockEvent = {
      key: 'Backspace'
    };

    input.props.onKeyUp(mockEvent);
  });

  it('handles keypress for Enter key', () => {
    const component = create(<Calculator />, {
      createNodeMock
    }).root;
    const input = component.findByProps({ tabIndex: -1 });
    const mockEvent = {
      key: 'Enter'
    };

    input.props.onKeyUp(mockEvent);
  });

  it('handles keypress for + key', () => {
    hooks.useCalculator.mockImplementation(() => ['0', setKey]);

    const component = create(<Calculator />, {
      createNodeMock
    }).root;
    const input = component.findByProps({ tabIndex: -1 });
    const mockEvent = {
      key: '+'
    };

    act(() => {
      input.props.onKeyUp(mockEvent);
    });

    expect(setKey).toHaveBeenCalledWith('+');
  });
});
