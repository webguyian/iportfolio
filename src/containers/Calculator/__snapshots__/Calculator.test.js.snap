// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> renders correctly 1`] = `
<div
  className="calculator-app"
>
  <output
    className="calculator-app-output calculator-app-output--1"
    onClick={[Function]}
  >
    0
  </output>
  <div
    className="calculator-app-input"
    onKeyUp={[Function]}
    tabIndex={-1}
  >
    <div
      className="calculator-app-row calculator-app-row--1"
    >
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        AC
        <span
          className="ui-text ui-text--accessible"
        >
          All clear
        </span>
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        +/-
        <span
          className="ui-text ui-text--accessible"
        >
          Positive/negative
        </span>
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        %
        <span
          className="ui-text ui-text--accessible"
        >
          Percent
        </span>
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        รท
        <span
          className="ui-text ui-text--accessible"
        >
          Divide
        </span>
      </button>
    </div>
    <div
      className="calculator-app-row calculator-app-row--2"
    >
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        7
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        8
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        9
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        ร
        <span
          className="ui-text ui-text--accessible"
        >
          Multiply
        </span>
      </button>
    </div>
    <div
      className="calculator-app-row calculator-app-row--3"
    >
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        4
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        5
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        6
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        -
        <span
          className="ui-text ui-text--accessible"
        >
          Subtract
        </span>
      </button>
    </div>
    <div
      className="calculator-app-row calculator-app-row--4"
    >
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        1
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        2
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        3
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        +
        <span
          className="ui-text ui-text--accessible"
        >
          Add
        </span>
      </button>
    </div>
    <div
      className="calculator-app-row calculator-app-row--5"
    >
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        0
        <span
          className="ui-text ui-text--accessible"
        />
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        .
        <span
          className="ui-text ui-text--accessible"
        >
          Decimal
        </span>
      </button>
      <button
        className="ui-btn ui-btn--rounded"
        onClick={[Function]}
        type="button"
      >
        =
        <span
          className="ui-text ui-text--accessible"
        >
          Equals
        </span>
      </button>
    </div>
  </div>
</div>
`;
